### PR TITLE
Add Analysis#teype_of test

### PR DESCRIPTION
**Pull request by h-michael**

### CODE REVIEW

File: crates/ra_analysis/tests/tests.rs [hunk 1]
```diff
@@ -2,14 +2,30 @@ use ra_syntax::TextRange;
...
     AnalysisChange, CrateGraph, FileId, FnSignatureInfo,
 };
 
 fn get_signature(text: &str) -> (FnSignatureInfo, Option<usize>) {
     let (analysis, position) = single_file_with_position(text);
     analysis.resolve_callable(position).unwrap().unwrap()
 }
+#[test]
+fn test_type_of() {
+    let (analysis, range) = analysis_and_range(
```
Comment (matklad) [inline crates/ra_analysis/tests/tests.rs]: existing `single_file_with_range` should be enough here.

File: crates/ra_analysis/tests/tests.rs [hunk 2]
```diff
@@ -2,14 +2,30 @@ use ra_syntax::TextRange;
...
+        "
+        //- /lib.rs
+        pub fn foo() -> u32 {
+            1
+        };
+
+        let <|>foo_test<|> = foo();
+    ",
```
Comment (matklad) [inline crates/ra_analysis/tests/tests.rs]: There are two reasons why this doesn't work.

* this is just an invalid syntax: the `let` is on the top-level, and one can't have a top-level `let` in Rust.
* `type_of` gets the type of the *expression* at the moment, it doesn't work for bindings. 

So, the following does work:

```
let (analysis, range) = single_file_with_range(
        "
        pub fn foo() -> u32 { 1 };

        fn main() {
            let foo_test = <|>foo()<|>;
        }
        ",
    );
```

File: crates/ra_analysis/tests/tests.rs (hunk 1 above)
Comment (h-michael) [reply crates/ra_analysis/tests/tests.rs]: Thanks, however, Should I remove `with_files_and_range` and `analysis_and_range` at this PR?

**Comment by h-michael** [comment]
@matklad 
Fixed.
I also added some cases that we can't use `type_of` now.

File: crates/ra_analysis/tests/type_of.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+fn test_type_of_for_num() {
+    let (analysis, range) = single_file_with_range(
+        r#"
+        fn main() {
+            let foo_test = <|>"foo"<|>;
+        }
+        "#,
+    );
+
+    analysis.type_of(range).unwrap().unwrap();
```
Comment (matklad) [inline crates/ra_analysis/tests/type_of.rs]: Let's replace the last `unwrap` with `assert!(.is_some())` here and elsewhere?

Comment (matklad) [reply crates/ra_analysis/tests/type_of.rs]: And we probably should add a `// FIXME: improve type_of to make this work`

**Comment by h-michael** [comment]
@matklad Fixed.

**Comment by matklad** [comment]
bors r+

Thanks!

**Pull request merged**